From 64b754a2ddf3363420c9afa7cee59e69b5c8a7f6 Mon Sep 17 00:00:00 2001
From: Mohammed Rafi KC <rkavunga@redhat.com>
Date: Wed, 23 Dec 2015 23:37:54 +0530
Subject: [PATCH 223/224] tier/unlink: symlink failed to unlink

during unlink of a file, we will get stat just after
deleting the file, to see if the file is under migration or not.

but this stat call will fail for symlink if the actual file
was deleted.

So it is better not to send stat request from client if it is
a symlink as we are not migrating symlink.

Back port of >
>Change-Id: Idc033b24fa3522b5261e579889d2195b43419682
>BUG: 1293963
>Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
>Reviewed-on: http://review.gluster.org/13074
>Tested-by: Gluster Build System <jenkins@build.gluster.com>
>Tested-by: NetBSD Build System <jenkins@build.gluster.org>
>Reviewed-by: Pranith Kumar Karampuri <pkarampu@redhat.com>
>Reviewed-by: Dan Lambright <dlambrig@redhat.com>
>Tested-by: Dan Lambright <dlambrig@redhat.com>

Change-Id: I19e37c18edd2870706bd58d64a840f8f6fbdbf34
BUG: 1293903
Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/64668
Reviewed-by: Dan Lambright <dlambrig@redhat.com>
Tested-by: Dan Lambright <dlambrig@redhat.com>
---
 xlators/cluster/dht/src/tier-common.c |    6 +++++-
 xlators/storage/posix/src/posix.c     |    3 ++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/xlators/cluster/dht/src/tier-common.c b/xlators/cluster/dht/src/tier-common.c
index 5a9f3ba..d8ea95e 100644
--- a/xlators/cluster/dht/src/tier-common.c
+++ b/xlators/cluster/dht/src/tier-common.c
@@ -553,12 +553,16 @@ tier_unlink (call_frame_t *frame, xlator_t *this, loc_t *loc, int xflag,
         }
 
         local->flags = xflag;
-        if (hashed_subvol == cached_subvol) {
+        if (IA_ISREG (loc->inode->ia_type) &&
+            (hashed_subvol == cached_subvol)) {
                 /*
                  * File resides in cold tier. We need to stat
                  * the file to see if it is being promoted.
                  * If yes we need to delete the destination
                  * file as well.
+                 *
+                 * Currently we are doing this check only for
+                 * regular files.
                  */
                 xdata = xdata ? dict_ref (xdata) : dict_new ();
                 if (xdata) {
diff --git a/xlators/storage/posix/src/posix.c b/xlators/storage/posix/src/posix.c
index 3eb2ea2..d19af48 100644
--- a/xlators/storage/posix/src/posix.c
+++ b/xlators/storage/posix/src/posix.c
@@ -1754,7 +1754,8 @@ posix_unlink (call_frame_t *frame, xlator_t *this,
                 goto out;
         }
 
-        if (xdata && dict_get (xdata, DHT_IATT_IN_XDATA_KEY)) {
+        if (IA_ISREG (loc->inode->ia_type) &&
+            xdata && dict_get (xdata, DHT_IATT_IN_XDATA_KEY)) {
                 fdstat_requested = 1;
         }
 
-- 
1.7.1

